components/greaselion/browser/greaselion_service_impl.cc

```diff
@@ -425,4 +432,40 @@ void GreaselionServiceImpl::SetBrowserVersionForTesting(
   UpdateInstalledExtensions();
 }
 
+void GreaselionServiceImpl::AddPrefListeners() {
+  // Currently we only listen for changes to the "Ads enabled" pref. If we add
+  // listeners for other relevant prefs we may want to debounce the extension
+  // update process in order to avoid intermediate, ephemeral extension sets.
+  pref_change_registrar_.Init(prefs_);
+  pref_change_registrar_.Add(
+      ads::prefs::kEnabled,
+      base::BindRepeating(&GreaselionServiceImpl::OnAnyPreferenceChanged,
+                          base::Unretained(this)));
+}
+
+void GreaselionServiceImpl::OnAnyPreferenceChanged() {
+  SetFeaturesUsingPrefs();
+  UpdateInstalledExtensions();
+}
+
+void GreaselionServiceImpl::SetFeaturesUsingPrefs() {
```

Comment:
we specifically did not want to do this with greaselion which is why we just have generic "features". Greaselion shouldn't have to know the details about every component that uses it.

Reply:
The existing dependency between the Rewards service and Greaselion causes issues where Greaselion is not in control of its own extension re-loading behavior. Although there are some different options for addressing that problem, removing the overly general and easily misused `SetFeature` API was the most straightforward way to solve it.

Are there use cases where these features will come from something that Greaselion does not have access to (e.g. Prefs or other profile-wide state)?

cc @pilgrim-brave

Reply:
The problem is that greaselion reloads all of the extensions when one feature is enabled/disabled and that's easy to fix. `SetFeature` is only easily misused because of that problem. This change forces GL to not only know about prefs for other components, in some cases it has to know what combination of prefs should enable or disable a feature. That definitely does not belong in GL imo and none of this solves the underlying issue that extensions are reloaded unnecessarily.

Reply:
> in some cases it has to know what combination of prefs should enable or disable a feature

That's right. It's a practical approach which works now and is easy to change later. I would give at least a 50% chance that this approach is sufficient for the foreseeable future. Also, if the use cases for GL are expanded, then I suspect that the currently implementation (design and all) will not scale well and will need to be reworked anyway.

> none of this solves the underlying issue that extensions are reloaded unnecessarily

In fact, it does solve the existing [double-load issue](https://github.com/brave/brave-browser/issues/21975) in its current form. As I noted in comments, any further work in GL will require solving at least the debouncing issue, and perhaps the "only reload what you need" problem. I judged that work to be out of scope for the problem I was trying to solve.